Userland/find.cpp

```diff
@@ -0,0 +1,468 @@
+                return false;
+            }
+            return WIFEXITED(status) && WEXITSTATUS(status) == 0;
+        }
+    }
+
+    Vector<char*> m_argv;
+};
+
+class AndCommand final : public Command {
+public:
+    AndCommand(NonnullOwnPtr<Command>&& lhs, NonnullOwnPtr<Command>&& rhs)
+        : m_lhs(move(lhs))
+        , m_rhs(move(rhs))
+    {
+    }
+
+private:
+    virtual bool evaluate(const char* file_path) const override
+    {
+        return m_lhs->evaluate(file_path) && m_rhs->evaluate(file_path);
+    }
+
+    NonnullOwnPtr<Command> m_lhs;
+    NonnullOwnPtr<Command> m_rhs;
+};
+
+class OrCommand final : public Command {
+public:
+    OrCommand(NonnullOwnPtr<Command>&& lhs, NonnullOwnPtr<Command>&& rhs)
+        : m_lhs(move(lhs))
+        , m_rhs(move(rhs))
+    {
+    }
+
+private:
+    virtual bool evaluate(const char* file_path) const override
+    {
+        return m_lhs->evaluate(file_path) || m_rhs->evaluate(file_path);
+    }
+
+    NonnullOwnPtr<Command> m_lhs;
+    NonnullOwnPtr<Command> m_rhs;
+};
+
+OwnPtr<Command> parse_complex_command(char* argv[]);
+
+// Parse a simple command starting at optind; leave optind at its the last
+// argument. Return nullptr if we reach the end of arguments.
+OwnPtr<Command> parse_simple_command(char* argv[])
+{
+    StringView arg = argv[optind];
+
+    if (arg.is_null()) {
+        return nullptr;
+    } else if (arg == "(") {
+        optind++;
+        auto command = parse_complex_command(argv);
+        if (command && argv[optind] && StringView(argv[++optind]) == ")")
+            return command;
+        fatal_error("Unmatched \033[1m(");
+    } else if (arg == "-type") {
+        return make<TypeCommand>(argv[++optind]);
+    } else if (arg == "-links") {
+        return make<LinksCommand>(argv[++optind]);
+    } else if (arg == "-user") {
+        return make<UserCommand>(argv[++optind]);
+    } else if (arg == "-group") {
+        return make<GroupCommand>(argv[++optind]);
+    } else if (arg == "-size") {
+        return make<SizeCommand>(argv[++optind]);
+    } else if (arg == "-print") {
+        g_have_seen_action_command = true;
+        return make<PrintCommand>();
+    } else if (arg == "-print0") {
+        g_have_seen_action_command = true;
+        return make<PrintCommand>(0);
+    } else if (arg == "-exec") {
+        g_have_seen_action_command = true;
+        Vector<char*> command_argv;
+        for (; argv[++optind] && StringView(argv[optind]) != ";";)
+            command_argv.append(argv[optind]);
+        return make<ExecCommand>(move(command_argv));
+    } else {
+        fatal_error("Unsupported command \033[1m%s", argv[optind]);
+    }
+}
+
+OwnPtr<Command> parse_complex_command(char* argv[])
+{
+    auto command = parse_simple_command(argv);
+
+    while (command && argv[optind] && argv[optind + 1]) {
+        StringView arg = argv[++optind];
+
+        enum { And,
+            Or } binary_operation
+            = And;
```

Comment:
What's up with the odd formatting here?

Reply:
clang-format ¯\_(ツ)_/¯

Manually changed to a single line for now.

Reply:
I think you can add a trailing comma to get clang-format to add an extra newline after the last element

Reply:
I don't like what it does with the first line either. I like either

```cpp
enum { And, Or } binary_operation = And;
```

or

```cpp
enum {
    And,
    Or,
} binary_operation = And;
```

Reply:
Adding a comma and bringing the And down manually will make clang-format produce the second one (I like that too)

Reply:
It produces

```cpp
enum {
    And,
    Or,
} binary_operation
    = And;
```

over here. (clang-format version 10.0.0)

Reply:
Ah, initialise with braces :)

Reply:
But I don't *want* to initialize with braces here, it's just a simple assignment.

Reply:
I guess you'll have to live with `// clang-format {off,on}` then

that idea of "braces for nontrivials" is a bit...backwards...
C++ would refuse to do some casts with braces that it would allow in `= ...`